airflow/dags/unzip_and_validate_gtfs_schedule/unzip_gtfs_schedule.py

```diff
@@ -0,0 +1,205 @@
+            content="\n".join(o.json() for o in self.outcomes).encode(),
+            exclude={"outcomes"},
+        )
+
+
+def summarize_zip_contents(zip: zipfile.ZipFile, at: str = ""):
+    files = []
+    directories = []
+    for entry in zip.namelist():
+        if zipfile.Path(zip, at=entry).is_file():
+            files.append(entry)
+        if zipfile.Path(zip, at=entry).is_dir():
+            directories.append(entry)
+    logging.info(f"Found files: {files} and directories: {directories}")
+    return files, directories
+
+
+def process_feed_files(
+    fs, extract: GTFSScheduleFeedExtract, zip: zipfile.ZipFile, directories: List[str]
+):
+    zipfile_files = []
+    # more than one directory --> can't parse
+    if len(directories) > 1:
+        raise ValueError("Unparseable zip: Multiple directories found")
+    # directory + any other artifact at top level --> can't parse
+    if directories and len([item for item in zipfile.Path(zip, at="").iterdir()]) > 1:
+        raise ValueError("Unparseable zip: Directory and other artifacts at root level")
+    # if the only artifact at the root level of the zipfile is a directory
+    # we can treat the contents of that directory as the feed
+    if [item.is_dir() for item in zipfile.Path(zip, at="").iterdir()] == [True]:
+        feed_directory = directories[0]
+        logging.info("Top-level directory found")
+    else:
+        feed_directory = ""
+
+    for file in zipfile.Path(zip, at=feed_directory).iterdir():
+        if feed_directory:
+            file_name = os.path.join(feed_directory, file.name)
+        else:
+            file_name = file.name
+        with zip.open(file_name) as f:
+            file_content = f.read()
+        file_extract = GTFSScheduleFeedFile(
+            ts=extract.ts,
+            base64_url=extract.base64_url,
+            zipfile_path=extract.path,
+            original_filename=file.name,
+            # TODO: do we need to make this safe?
+            filename=file.name,
+        )
+        file_extract.save_content(content=file_content, fs=fs)
+        zipfile_files.append(file_extract)
+    return zipfile_files
+
+
+def unzip_individual_feed(
+    fs, extract: GTFSScheduleFeedExtract
+) -> GTFSScheduleFeedExtractUnzipOutcome:
+    """If zip contains a directory and nothing else at top level or files at top level,
+    we return the contents of the top level or the unique directory as the feed.
+    If there's a directory *and* other files at top level within zip, or if
+    there are nested directories, say we can't parse."""
+    logging.info(f"Processing {extract.name}")
+    zipfile_md5_hash = ""
+    try:
+        with fs.open(extract.path) as f:
+            zipfile_md5_hash = f.info()["md5Hash"]
+            zip = zipfile.ZipFile(f)
+            files, directories = summarize_zip_contents(zip)
+            try:
+                zipfile_files = process_feed_files(fs, extract, zip, directories)
+            except ValueError as e:
```

Comment:
If you wanted you could just catch the ValueError in an additional except of the first try, to avoid some nesting